homeassistant/helpers/event.py

```diff
@@ -90,34 +99,246 @@ def state_change_listener(event):
+    string result, such as 'On', 'Open', or 'Yes'. If the template results
+    in an error state when the value changes, this will be logged and not
+    passed through.
+
+    On template registration, the action will be called with an entity_id
+    and state of some entity referenced by the template, or '*' if no
+    entity is referenced by the template.
+
+    If the initial run of the template is invalid and results in an
+    exception, the listener will still be registered but will only
+    fire if the template result becomes true without an exception.
+
+    Action arguments
+    ----------------
+    entity_id
+        ID of the entity that triggered the state change.
+    old_state
+        The old state of the entity that changed.
+    new_state
+        New state of the entity that changed.
+
+    Parameters
+    ----------
+    hass
+        Home assistant object.
+    template
+        The template to calculate.
+    action
+        Callable to call with results. See above for arguments.
+    variables
+        Variables to pass to the template.
+
+    Returns
+    -------
+    Callable to unregister the listener.
+
+    """
+    @callback
+    def state_changed_listener(event, template, last_result, result):
+        """Check if condition is correct and run action."""
+        if isinstance(result, TemplateError):
+            _LOGGER.exception(result)
+            return
+
+        if last_result is not None:
+            if not _boolean_coerce(last_result) \
+                    and _boolean_coerce(result):
+                hass.async_run_job(action, event.data.get('entity_id'),
+                                   event.data.get('old_state'),
+                                   event.data.get('new_state'))
+        elif _boolean_coerce(result):
+            # First run of the listener. Figure out an entity ID to
+            # pass back to the action because it expects one.
+            info = template.async_render_to_info(variables)
+            state = None
+            entity_id = None
+            #  pylint: disable=protected-access
+            for eid in info._entities:
+                state = hass.states.get(eid)
+                if state:
+                    entity_id = eid
+                    break
+            if not state:
+                for st in hass.states.async_all():
+                    if info.filter_lifecycle(st.entity_id):
+                        state = st
+                        entity_id = st.entity_id
+                        break
+            hass.async_run_job(
+                action, entity_id or MATCH_ALL, state, state)
+
+    info = async_track_template_result(
+        hass, template, state_changed_listener, variables)
+
+    return info.async_remove
+
+
+track_template = threaded_listener_factory(async_track_template)
+
+
+_UNCHANGED = object()
+
+
+class TrackTemplateResultInfo:
+    """Return value from async_track_template_result."""
```

Comment:
We're doing more than just returning a value, as we're also running an action?

Reply:
This gets handed back so there's a place for async_refresh, for when the template entity using it gets manually refreshed its got something to call.

Reply:
I meant that the doc string does not reflect the duties of this class.